### PR TITLE
Ensure that the result for PowMod is fully initialized

### LINKED ISSUE #2575: [perf pipeline] `Perf_BigInteger.ModPow` failing with `System.ArgumentException: Destination is too short. (Parameter 'destination')`
[Log from runtime-wasm-perf](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-74097-merge-452b23c3e6b4475d86/wasm.x64.micro.net7.0.Partition5/1/console.8998d92c.log?helixlogtype=result) run on https://github.com/dotnet/runtime/pull/74097:

```
// **************************
// Benchmark: Perf_BigInteger.ModPow: Job-BJXMHI(PowerPlanMode=00000000-0000-0000-0000-000000000000, Runtime=Wasm, Toolchain=Wasm, InvocationCount=1, IterationCount=1, IterationTime=250.0000 ms, MaxIterationCount=20, MinIterationCount=15, RunStrategy=ColdStart, UnrollFactor=1, WarmupCount=0) [arguments=1024,1024,64 bits]
// *** Execute ***
// Launch: 1 / 1
// Execute: /home/helixbot/.jsvu/v8 --expose_wasm test-main.js -- --run 95a3fee8-6d92-4c8b-8059-369b73cd756f.dll --benchmarkName "System.Numerics.Tests.Perf_BigInteger.ModPow(arguments: 1024,1024,64 bits)" --job "PowerPlanMode=00000000-0000-0000-0000-000000000000, Runtime=Wasm, Toolchain=Wasm, InvocationCount=1, IterationCount=1, IterationTime=250.0000 ms, MaxIterationCount=20, MinIterationCount=15, RunStrategy=ColdStart, UnrollFactor=1, WarmupCount=0" --benchmarkId 67  in /datadisks/disk1/work/A6C10953/w/B67009A6/e/performance/artifacts/bin/for-running/MicroBenchmarks/95a3fee8-6d92-4c8b-8059-369b73cd756f/bin/net7.0/browser-wasm/AppBundle
Failed to set up high priority. Make sure you have the right permissions. Message: Permission denied
Incoming arguments: --run 95a3fee8-6d92-4c8b-8059-369b73cd756f.dll --benchmarkName System.Numerics.Tests.Perf_BigInteger.ModPow(arguments: 1024,1024,64 bits) --job PowerPlanMode=00000000-0000-0000-0000-000000000000, Runtime=Wasm, Toolchain=Wasm, InvocationCount=1, IterationCount=1, IterationTime=250.0000 ms, MaxIterationCount=20, MinIterationCount=15, RunStrategy=ColdStart, UnrollFactor=1, WarmupCount=0 --benchmarkId 67
Application arguments: --run 95a3fee8-6d92-4c8b-8059-369b73cd756f.dll --benchmarkName System.Numerics.Tests.Perf_BigInteger.ModPow(arguments: 1024,1024,64 bits) --job PowerPlanMode=00000000-0000-0000-0000-000000000000, Runtime=Wasm, Toolchain=Wasm, InvocationCount=1, IterationCount=1, IterationTime=250.0000 ms, MaxIterationCount=20, MinIterationCount=15, RunStrategy=ColdStart, UnrollFactor=1, WarmupCount=0 --benchmarkId 67
console.debug: mono_wasm_runtime_ready fe00e07a-5519-4dfe-b35a-f867dbaf2e28
console.info: Initializing.....
// BeforeAnythingElse

// Benchmark Process Environment Information:
// Runtime=.NET Core (Mono) 8.0.0-ci, Wasm NativeAOT
// GC=Non-concurrent Workstation
// HardwareIntrinsics=
// Job: Job-WFMUPX(PowerPlanMode=00000000-0000-0000-0000-000000000000, InvocationCount=1, IterationCount=1, IterationTime=250.0000 ms, MaxIterationCount=20, MinIterationCount=15, RunStrategy=ColdStart, UnrollFactor=1, WarmupCount=0)

// BeforeActualRun
```
...
```
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> System.ArgumentException: Destination is too short. (Parameter 'destination')
   at System.Numerics.BigIntegerCalculator.PowCore(Span`1 , Int32 , ReadOnlySpan`1 , ReadOnlySpan`1 , Span`1 , Span`1 )
   at System.Numerics.BigIntegerCalculator.Pow(ReadOnlySpan`1 , ReadOnlySpan`1 , ReadOnlySpan`1 , Span`1 )
   at System.Numerics.BigInteger.ModPow(BigInteger , BigInteger , BigInteger )
   at System.Numerics.Tests.Perf_BigInteger.ModPow(BigIntegers arguments)
   at BenchmarkDotNet.Autogenerated.Runnable_67.WorkloadActionUnroll(Int64 invokeCount)
   at BenchmarkDotNet.Engines.Engine.RunIteration(IterationData data)
   at BenchmarkDotNet.Engines.EngineStage.RunIteration(IterationMode mode, IterationStage stage, Int32 index, Int64 invokeCount, Int32 unrollFactor)
   at BenchmarkDotNet.Engines.EngineActualStage.RunSpecific(Int64 invokeCount, IterationMode iterationMode, Int32 iterationCount, Int32 unrollFactor)
   at BenchmarkDotNet.Engines.EngineActualStage.Run(Int64 invokeCount, IterationMode iterationMode, Boolean runAuto, Int32 unrollFactor, Boolean forceSpecific)
   at BenchmarkDotNet.Engines.EngineActualStage.RunWorkload(Int64 invokeCount, Int32 unrollFactor, Boolean forceSpecific)
   at BenchmarkDotNet.Engines.Engine.Run()
   at BenchmarkDotNet.Autogenerated.Runnable_67.Run(IHost host, String benchmarkName)
   at System.Reflection.MethodInvoker.InterpretedInvoke(Object , Span`1 , BindingFlags )
   --- End of inner exception stack trace ---
   at System.Reflection.RuntimeMethodInfo.Invoke(Object , BindingFlags , Binder , Object[] , CultureInfo )
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at BenchmarkDotNet.Autogenerated.UniqueProgramName.AfterAssemblyLoadingAttached(String[] args)
// AfterAll
{"name":"ExitStatus","message":"Program terminated with exit(-1)","status":-1}
No Workload Results were obtained from the run.
// Benchmark Process 4941 has exited with code 255.
```

This is reproducible on main, with interpreter.

cc @adamsitnik @LoopedBard3 @lewing

### PR DESCRIPTION
**Pull request by tannergooding**
This resolves dotnet/performance#2575

The fix provided by https://github.com/dotnet/runtime/pull/74112 was not quite correct. It attempted to resolve the issue by no longer slicing the intermediate `result` to the computed `resultLength`.

While this did fix the immediate issue, it exposed a new issue where the intermediate `result` buffer could then be larger than the destination `bits` (the final result) buffer.

The root issue for https://github.com/dotnet/runtime/pull/74112 was that while `bits` is the final result buffer, it was also being used as a temporary storage location and thus could have values at indices greater than `resultLength` which were non-zero.

There are multiple ways this could be fixed, but I believe the least risky is to continue slicing `result` down to `resultLength` (effectively reverting the fix from https://github.com/dotnet/runtime/pull/74112) and to instead explicitly clear any indices from `bits` that are greater than `result.Length`.

Assuming that the computed `resultLength` is correct (and it should be or we have more serious issues), then this ensures that `bits` is exactly `result` and all remaining indices are zero, which should ensure a correct result.

-----------------------

Alternatively we could continue not slicing `result` where `resultLength` was being computed and instead do something like:
```csharp
Span<uint> result = PowCore(valueCopy, value.Length, temp, power, bits);
int resultLength = Math.Min(result.Length, bits.Length);
result.CopyTo(bits);
```

However, this might imply that the computed `resultLength` can't be relied upon and could mask an issue in some future refactoring.

-or-

We could add an additional allocation for the temporary data rather than using `bits`, which is an overall "better" change (less error prone and safer; but also more "expensive"). However, this would also be more involved and therefore riskier to take into RC2.

### CODE REVIEW

**Comment by tannergooding** [comment]
CC. @dakersnar, @sakno

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.PowMod.cs [hunk 1]
```diff
@@ -50,19 +52,19 @@ private static Span<uint> PowCore(Span<uint> value, int valueLength, Span<uint>
...
                 if ((power & 1) == 1)
-                    bitsLength = MultiplySelf(ref result, bitsLength, value.Slice(0, valueLength), ref temp);
+                    resultLength = MultiplySelf(ref result, resultLength, value.Slice(0, valueLength), ref temp);
                 if (power != 1)
                     valueLength = SquareSelf(ref value, valueLength, ref temp);
                 power >>= 1;
             }
 
-            return result;
+            return result.Slice(0, resultLength);
```
Comment (dakersnar) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.PowMod.cs:67]: Notably, this function and functions that use it were not touched by https://github.com/dotnet/runtime/pull/74112. However, if you think the same bug was already present here before that PR was merged, then this is fine to include.

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.PowMod.cs [hunk 2]
```diff
@@ -531,7 +541,7 @@ private const
...
```
Comment (dakersnar) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.PowMod.cs:544]: Same here, this was not changed by https://github.com/dotnet/runtime/pull/74112.

**Review by dakersnar — APPROVED** [review]
Looks good to me, thanks for the fix. Glad we caught this.

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.PowMod.cs (hunk 1 above)
Comment (tannergooding) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.PowMod.cs:67]: The concern is that a similar bug could exist. The computed `resultLength` should be correct and so slicing should not be harmful and will help avoid similar issues as to what we already saw.

**Pull request merged**